core/src/main/java/org/apache/iceberg/puffin/PuffinWriter.java

```diff
@@ -0,0 +1,189 @@
+      ByteBuffer rawData = PuffinFormat.compress(codec, blob.blobData());
+      int length = rawData.remaining();
+      writeFully(rawData);
+      writtenBlobsMetadata.add(new BlobMetadata(blob.type(), blob.inputFields(), fileOffset, length,
+          codec.codecName(), blob.properties()));
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+  }
+
+  @Override
+  public Metrics metrics() {
+    return new Metrics();
+  }
+
+  @Override
+  public long length() {
+    return fileSize();
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (!finished) {
+      finish();
+    }
+
+    outputStream.close();
+  }
+
+  private void writeHeaderIfNeeded() throws IOException {
+    if (headerWritten) {
+      return;
+    }
+
+    this.outputStream.write(MAGIC);
+    this.headerWritten = true;
+  }
+
+  public void finish() throws IOException {
+    checkNotFinished();
+    writeHeaderIfNeeded();
+    if (footerSize.isPresent()) {
+      throw new IllegalStateException("footerSize already set");
+    }
+
+    long footerOffset = outputStream.getPos();
+    writeFooter();
+    this.footerSize = Optional.of(Math.toIntExact(outputStream.getPos() - footerOffset));
+    this.fileSize = Optional.of(outputStream.getPos());
+    this.finished = true;
+  }
+
+  private void writeFooter() throws IOException {
+    FileMetadata fileMetadata = new FileMetadata(writtenBlobsMetadata, properties);
+    ByteBuffer footerJson = ByteBuffer.wrap(FileMetadataParser.toJson(fileMetadata, false)
+        .getBytes(StandardCharsets.UTF_8));
+    ByteBuffer footerPayload = PuffinFormat.compress(footerCompression, footerJson);
+    outputStream.write(MAGIC);
+    int footerPayloadLength = footerPayload.remaining();
+    writeFully(footerPayload);
+    PuffinFormat.writeIntegerLittleEndian(outputStream, footerPayloadLength);
+    writeFlags();
+    outputStream.write(MAGIC);
+  }
+
+  private void writeFlags() throws IOException {
+    Map<Integer, List<Flag>> flagsByByteNumber = fileFlags().stream()
+        .collect(Collectors.groupingBy(Flag::byteNumber));
+    for (int byteNumber = 0; byteNumber < PuffinFormat.FOOTER_STRUCT_FLAGS_LENGTH; byteNumber++) {
+      int byteFlag = 0;
+      for (Flag flag : flagsByByteNumber.getOrDefault(byteNumber, ImmutableList.of())) {
+        byteFlag |= 0x1 << flag.bitNumber();
+      }
+      outputStream.write(byteFlag);
+    }
+  }
+
+  private void writeFully(ByteBuffer buffer) throws IOException {
+    WritableByteChannel channel = Channels.newChannel(outputStream);
+    while (buffer.remaining() > 0) {
+      channel.write(buffer);
+    }
+  }
+
+  public long footerSize() {
+    return footerSize.orElseThrow(() -> new IllegalStateException("Footer not written yet"));
+  }
+
+  public long fileSize() {
+    return fileSize.orElseThrow(() -> new IllegalStateException("File not written yet"));
```

Comment:
`written` -> `closed`?

Reply:
no, because it's written (but not closed) in finish